File: 013_Binary_Search_Interview_Questions/First_and_Last_Occurrence.py
def firstOcc(arr, n, key):

   start = 0
   end = n - 1
   mid = start + (end - start) // 2
   ans = -1

   while start <= end:
       if arr[mid] == key:
           ans = mid
           end = mid - 1  # search for an earlier occurrence
       elif key > arr[mid]:
           start = mid + 1  # search in the right subarray
       else:
           end = mid - 1  # search in the left subarray
       mid = start + (end - start) // 2

   return ans


def lastOcc(arr, n, key):

   start = 0
   end = n - 1
   mid = start + (end - start) // 2
   ans = -1

   while start <= end:
       if arr[mid] == key:
           ans = mid
           start = mid + 1  # search for a later occurrence
       elif key > arr[mid]:
           start = mid + 1  # search in the right subarray
       else:
           end = mid - 1  # search in the left subarray
       mid = start + (end - start) // 2

   return ans


odd = [1, 2, 3, 7, 7, 7, 7, 8, 9]
print("First occurrence of 7 is", firstOcc(odd, 9, 7))
print("Last occurrence of 7 is", lastOcc(odd, 9, 7))

File: 013_Binary_Search_Interview_Questions/First_and_Last_Occurrence.swift

var arr: [Int] = [1, 8, 8, 8, 9, 10]

func firstOccurrence(arr: [Int], key: Int) -> Int {

    var start = 0;
    var end = arr.count - 1;
    var mid = start + (end - start) / 2;
    var ans: Int = 0

  while start <= end {
    if(arr[mid] == key){
      ans = mid
      end = mid - 1
    }
    else if(arr[mid] < key){
      start = mid+1
    }
    else{
      end = mid - 1
    }

    mid = start + (end - start)/2


  }
  return ans

}  

func lastOccurrence(arr: [Int], key: Int) -> Int {

    var start = 0;
    var end = arr.count - 1;
    var mid = start + (end - start) / 2;
    var ans: Int = 0

   while start <= end {

        if(arr[mid] == key){
            ans = mid
            start = mid+1
        }
        else if(arr[mid] < key){
            start = mid+1
        }
        else if(arr[mid]>key){
            end = mid - 1
        }
        mid = start + (end - start)/2

  }
  return ans
}

print(firstOccurrence(arr: arr, key: 8))
print(lastOccurrence(arr: arr, key: 8))